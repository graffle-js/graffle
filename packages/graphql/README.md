# @graffle/graphql

Generic GraphQL utilities for schema analysis, document operations, and type definitions.

This package provides tree-shakeable, reusable GraphQL tools that can be used standalone or as part of the Graffle ecosystem.

## Features

- **Schema Utilities**: Schema introspection, analysis, and type operations
- **Document Utilities**: GraphQL document parsing and manipulation
- **HTTP Helpers**: HTTP request/response utilities for GraphQL
- **Execution Helpers**: GraphQL execution utilities
- **Type Definitions**: Complete TypeScript type definitions for GraphQL schemas

## Installation

```bash
npm install @graffle/graphql graphql
# or
pnpm add @graffle/graphql graphql
```

## Usage

```typescript
import * as Graphql from '@graffle/graphql'
import * as Schema from '@graffle/graphql/schema'
import * as Document from '@graffle/graphql/document'
```

## License

MIT
