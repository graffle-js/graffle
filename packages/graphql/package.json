{
  "name": "@graffle/graphql",
  "version": "0.1.0",
  "description": "Generic GraphQL utilities for schema analysis, document operations, and type definitions",
  "type": "module",
  "exports": {
    ".": "./build/index.js",
    "./document": "./build/document.js",
    "./document-types": "./build/document/__.js",
    "./schema": "./build/schema/index.js",
    "./http": "./build/http.js",
    "./execute": "./build/execute.js",
    "./schema-types": "./build/schema-types/index.js",
    "./_Nodes.js": "./build/_Nodes.js",
    "./typed-document/_.js": "./build/typed-document/_.js"
  },
  "imports": {
    "#~/*": "./build/*"
  },
  "files": [
    "build",
    "src"
  ],
  "scripts": {
    "build": "tsc --build tsconfig.build.json",
    "build:emit": "tsc --build tsconfig.build.json --noCheck",
    "build:clean": "tsc --build tsconfig.build.json --clean && rm -rf build",
    "dev": "tsc --build tsconfig.build.json --watch --noCheck",
    "check": "pnpm run '/check:.*/'",
    "check:types": "tsc --noEmit",
    "test": "vitest",
    "test:unit": "vitest --run"
  },
  "dependencies": {
    "@0no-co/graphql.web": "1.2.0",
    "@wollybeard/kit": "^0.98.0"
  },
  "peerDependencies": {
    "graphql": "14 - 16"
  },
  "devDependencies": {
    "@test/fixtures": "workspace:*"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/graffle-js/graffle.git",
    "directory": "packages/graphql"
  },
  "keywords": [
    "graphql",
    "schema",
    "document",
    "utilities"
  ],
  "author": {
    "name": "Jason Kuhrt",
    "email": "jasonkuhrt@me.com",
    "url": "https://kuhrt.me"
  },
  "license": "MIT"
}
