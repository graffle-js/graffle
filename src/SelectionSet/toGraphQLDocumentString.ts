import type { SelectionSet } from './__.js'
import { aliasPattern, fragmentPattern } from './SelectionSet.js'

type SpecialFields = {
  // todo - this requires having the schema at runtime to know which fields to select.
  // $scalars?: SelectionSet.Indicator
  $include?: SelectionSet.Directive.Include['$include']
  $skip?: SelectionSet.Directive.Skip['$skip']
  $defer?: SelectionSet.Directive.Defer['$defer']
  $stream?: SelectionSet.Directive.Stream['$stream']
  $?: Args
}

type Args = { [k: string]: Args_ }

type Args_ = string | boolean | null | number | Args

type Indicator = 0 | 1 | boolean

export type GraphQLDocumentObject = {
  [k: string]: Indicator | SS
}

type SS = {
  [k: string]: Indicator | SS
} & SpecialFields

export const toGraphQLDocumentString = (ss: GraphQLDocumentObject) => {
  let docString = ``
  docString += `query {
		${selectionSet(ss)}
	}`
  return docString
}

const directiveArgs = (config: object) => {
  return Object.entries(config).filter(([_, v]) => v !== undefined).map(([k, v]) => {
    return `${k}: ${JSON.stringify(v)}`
  }).join(`, `)
}

const indicatorOrSelectionSet = (ss: null | Indicator | SS): string => {
  if (ss === null) return `null` // todo test this case
  if (isIndicator(ss)) return ``

  const { $include, $skip, $defer, $stream, $, ...rest } = ss

  let args = ``
  let directives = ``

  if ($stream !== undefined) {
    const config = {
      if: typeof $stream === `boolean` ? $stream : $stream.if === undefined ? true : $stream.if,
      label: typeof $stream === `boolean` ? undefined : $stream.label,
      initialCount: typeof $stream === `boolean` ? undefined : $stream.initialCount,
    }
    directives += `@stream(${directiveArgs(config)})`
  }

  if ($defer !== undefined) {
    const config = {
      if: typeof $defer === `boolean` ? $defer : $defer.if === undefined ? true : $defer.if,
      label: typeof $defer === `boolean` ? undefined : $defer.label,
    }
    directives += `@defer(${directiveArgs(config)})`
  }

  if ($include !== undefined) {
    directives += `@include(if: ${
      String(typeof $include === `boolean` ? $include : $include.if === undefined ? true : $include.if)
    })`
  }

  if ($skip !== undefined) {
    directives += `@skip(if: ${String(typeof $skip === `boolean` ? $skip : $skip.if === undefined ? true : $skip.if)})`
  }

  if ($ !== undefined) {
    const entries = Object.entries($)
    args = entries.length === 0 ? `` : `(${
      entries.map(([k, v]) => {
        return `${k}: ${JSON.stringify(v)}`
      }).join(`, `)
    })`
  }

  if (Object.keys(rest).length === 0) {
    return `${args} ${directives}`
  }

  return `${args} ${directives} {
		${selectionSet(rest)}
	}`
}

const selectionSet = (ss: GraphQLDocumentObject) => {
  return Object.entries(ss).filter(([_, v]) => {
    return isPositiveIndicator(v)
  }).map(([field, ss]) => {
    return `${resolveFragment(resolveAlias(field))} ${indicatorOrSelectionSet(ss)}`
  }).join(`\n`) + `\n`
}

// todo use a given schema to ensure that field is actually a fragment and not just happened to be using pattern onX
const resolveFragment = (field: string) => {
  const match = field.match(fragmentPattern)
  if (match?.groups) {
    return `...on ${match.groups[`name`]!}`
  }
  return field
}

// todo use a given schema to ensure that field is actually a fragment and not just happened to be using pattern onX
const resolveAlias = (field: string) => {
  const match = field.match(aliasPattern)
  if (match?.groups) {
    return `${match.groups[`actual`]!}: ${match.groups[`alias`]!}`
  }
  return field
}

const isIndicator = (v: any): v is Indicator => {
  return String(v) in indicator
}

const isPositiveIndicator = (v: any): v is SelectionSet.ClientIndicatorPositive => {
  return !(String(v) in negativeIndicator)
}

const negativeIndicator = {
  '0': 0,
  'false': false,
  'undefined': undefined,
}

const positiveIndicator = {
  '1': 1,
  'true': true,
}

const indicator = {
  ...negativeIndicator,
  ...positiveIndicator,
}
