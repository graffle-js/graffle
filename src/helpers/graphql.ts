/**
 * Clean a GraphQL document to send it via a GET query
 */
export const cleanQuery = (str: string): string => str.replace(/([\s,]|#[^\n\r]+)+/g, ` `).trim()
