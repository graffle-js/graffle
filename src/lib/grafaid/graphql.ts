import type { GraphQLEnumValue, GraphQLError, GraphQLField, GraphQLInputField, GraphQLNamedType } from 'graphql'
import {
  GraphQLEnumType,
  GraphQLInputObjectType,
  GraphQLInterfaceType,
  GraphQLList,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLScalarType,
  GraphQLUnionType,
  isEnumType,
  isInputObjectType,
  isInterfaceType,
  isObjectType,
  isScalarType,
  isUnionType,
  OperationTypeNode,
} from 'graphql'
import type { Errors } from '../errors/__.js'
import { isString } from '../prelude.js'
import { Nodes } from './_Nodes.js'
import { TypedDocument } from './typed-document/__.js'

export * from './_Nodes.js'

export const StandardScalarTypeNames = {
  String: `String`,
  ID: `ID`,
  Int: `Int`,
  Float: `Float`,
  Boolean: `Boolean`,
}

export type StandardScalarTypeNames = keyof typeof StandardScalarTypeNames

const TypeScriptPrimitiveTypeNames = {
  string: `string`,
  number: `number`,
  boolean: `boolean`,
}
type TypeScriptPrimitiveTypeNames = keyof typeof TypeScriptPrimitiveTypeNames

export const StandardScalarTypeTypeScriptMapping = {
  String: `string`,
  ID: `string`,
  Int: `number`,
  Float: `number`,
  Boolean: `boolean`,
} satisfies Record<
  StandardScalarTypeNames,
  TypeScriptPrimitiveTypeNames
>

export type TypeNamedKind = `Enum` | `InputObject` | `Interface` | `Object` | `Scalar` | `Union`

export type TypeMapKind = TypeNamedKind | `Root`

export const operationTypeNameToRootTypeName = {
  query: `Query`,
  mutation: `Mutation`,
  subscription: `Subscription`,
} as const

export const RootTypeNameToOperationName = {
  Query: OperationTypeNode.QUERY,
  Mutation: OperationTypeNode.MUTATION,
  Subscription: OperationTypeNode.SUBSCRIPTION,
} as const

export type RootTypeNameToOperationName = typeof RootTypeNameToOperationName

export const isStandardScalarType = (type: GraphQLScalarType) => {
  return type.name in StandardScalarTypeNames
}

export type ClassToName<C> = C extends GraphQLScalarType ? `GraphQLScalarType`
  : C extends GraphQLObjectType ? `GraphQLObjectType`
  : C extends GraphQLInterfaceType ? `GraphQLInterfaceType`
  : C extends GraphQLUnionType ? `GraphQLUnionType`
  : C extends GraphQLEnumType ? `GraphQLEnumType`
  : C extends GraphQLInputObjectType ? `GraphQLInputObjectType`
  : C extends GraphQLList<any> ? `GraphQLList`
  : C extends GraphQLNonNull<any> ? `GraphQLNonNull`
  : never

export const NamedNameToClass = {
  GraphQLScalarType: GraphQLScalarType,
  GraphQLObjectType: GraphQLObjectType,
  GraphQLInterfaceType: GraphQLInterfaceType,
  GraphQLUnionType: GraphQLUnionType,
  GraphQLEnumType: GraphQLEnumType,
  GraphQLInputObjectType: GraphQLInputObjectType,
} as const

export type NamedNameToClass = typeof NamedNameToClass

export const NameToClass = {
  GraphQLNonNull: GraphQLNonNull,
  GraphQLList: GraphQLList,
  ...NamedNameToClass,
} as const

export type AnyGraphQLOutputField = GraphQLField<any, any>

export type AnyField = AnyGraphQLOutputField | GraphQLInputField

export type NameToClass = typeof NameToClass

export type NodeName = keyof NameToClass

export type NodeNamePlus = NodeName | 'GraphQLField'

export type AnyNamedClassName = keyof NamedNameToClass

export type AnyClass = InstanceType<NameToClass[keyof NameToClass]>

/**
 * Groups
 */

export type Describable =
  | GraphQLNamedType
  | AnyField

export const getTypeNameAndKind = (
  node: GraphQLNamedType,
): { name: string; kind: 'Object' | 'Interface' | 'Union' | 'Enum' | 'Scalar' } => {
  const name = node.name
  const kind = getNodeKindOld(node).replace(`GraphQL`, ``).replace(`Type`, ``) as
    | 'Object'
    | 'Interface'
    | 'Union'
    | 'Enum'
    | 'Scalar'
  return { name, kind }
}

export const getTypeKind = <$Node extends GraphQLNamedType>(node: $Node): ClassToName<$Node> => {
  switch (true) {
    case isObjectType(node):
      return `GraphQLObjectType` as ClassToName<$Node>
    case isInputObjectType(node):
      return `GraphQLInputObjectType` as ClassToName<$Node>
    case isUnionType(node):
      return `GraphQLUnionType` as ClassToName<$Node>
    case isInterfaceType(node):
      return `GraphQLInterfaceType` as ClassToName<$Node>
    case isEnumType(node):
      return `GraphQLEnumType` as ClassToName<$Node>
    case isScalarType(node):
      return `GraphQLScalarType` as ClassToName<$Node>
    default:
      throw new Error(`Unknown node kind: ${String(node)}`)
  }
}

export const getNodeKindOld = (node: Describable): NodeNamePlus => {
  switch (true) {
    case isObjectType(node):
      return `GraphQLObjectType`
    case isInputObjectType(node):
      return `GraphQLInputObjectType`
    case isUnionType(node):
      return `GraphQLUnionType`
    case isInterfaceType(node):
      return `GraphQLInterfaceType`
    case isEnumType(node):
      return `GraphQLEnumType`
    case isScalarType(node):
      return `GraphQLScalarType`
    default:
      return `GraphQLField`
  }
}

export const getNodeDisplayName = (node: Describable) => {
  return toDisplayName(getNodeKindOld(node))
}

const toDisplayName = (nodeName: NodeNamePlus) => {
  return nodeName.replace(/^GraphQL/, ``).replace(/Type$/, ``)
}

export const isDeprecatableNode = (node: object): node is GraphQLEnumValue | AnyField => {
  return `deprecationReason` in node
}

export interface RequestInput {
  query: string | TypedDocument.TypedDocument
  variables?: Variables
  operationName?: string
}

export type Variables = {
  [key: string]: string | boolean | null | number | Variables
}

export type SomeData = Record<string, any>

export type GraphQLExecutionResultError = Errors.ContextualAggregateError<GraphQLError>

const definedOperationPattern = new RegExp(`^\\b(${Object.values(OperationTypeNode).join(`|`)})\\b`)

export const parseOperationType = (request: RequestInput): OperationTypeNode | null => {
  const { operationName, query: document } = request

  const documentUntyped = TypedDocument.unType(document)

  if (!isString(documentUntyped)) {
    for (const node of documentUntyped.definitions) {
      if (node.kind === Nodes.Kind.OPERATION_DEFINITION) {
        if (operationName ? node.name?.value === operationName : true) {
          return node.operation
        }
      }
    }
    throw new Error(`Could not parse operation type from document.`)
  }

  const definedOperations = documentUntyped.split(/[{}\n]+/).map(s => s.trim()).map(line => {
    const match = line.match(definedOperationPattern)
    if (!match) return null
    return {
      line,
      operationType: match[0] as OperationTypeNode,
    }
  }).filter(_ => _ !== null)
  // console.log(definedOperations)

  // Handle obviously invalid cases that are zero cost to compute.

  // The given operation name will not match to anything.
  if (definedOperations.length > 1 && !request.operationName) return null

  // An operation name is required but was not given.
  if (definedOperations.length === 0 && request.operationName) return null

  // Handle optimistically assumed valid case short circuits.

  if (definedOperations.length === 0) {
    // Assume that the implicit query syntax is being used.
    // This is a non-validated optimistic approach for performance, not aimed at correctness.
    // For example its not checked if the document is actually of the syntactic form `{ ... }`
    return OperationTypeNode.QUERY
  }

  // Continue to the full search.

  const definedOperationToAnalyze = operationName
    ? definedOperations.find(o => o.line.includes(operationName))
    : definedOperations[0]

  // Invalid: The given operation name does not show up in the document.
  if (!definedOperationToAnalyze) return null

  return definedOperationToAnalyze.operationType
}
