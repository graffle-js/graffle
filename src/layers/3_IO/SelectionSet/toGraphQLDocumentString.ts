import { RootTypeName } from '../../../lib/graphql.js'
import { lowerCaseFirstLetter } from '../../../lib/prelude.js'
import { Schema } from '../../1_Schema/__.js'
import { readMaybeThunk } from '../../1_Schema/core/helpers.js'
import type { ReturnModeType } from '../../4_client/Config.js'
import type { SelectionSet } from './__.js'
import { aliasPattern, fragmentPattern } from './SelectionSet.js'

type SpecialFields = {
  // todo - this requires having the schema at runtime to know which fields to select.
  // $scalars?: SelectionSet.Indicator
  $include?: SelectionSet.Directive.Include['$include']
  $skip?: SelectionSet.Directive.Skip['$skip']
  $defer?: SelectionSet.Directive.Defer['$defer']
  $stream?: SelectionSet.Directive.Stream['$stream']
  $?: Args
}

type Args = { [k: string]: Args_ }

type Args_ = string | boolean | null | number | Args

type Indicator = 0 | 1 | boolean

export type DocumentObject = Record<string, GraphQLRootSelection>

export type GraphQLRootSelection = { query: GraphQLObjectSelection } | { mutation: GraphQLObjectSelection }

export type GraphQLObjectSelection = Record<string, Indicator | SS>

export type SS = {
  [k: string]: Indicator | SS
} & SpecialFields

export interface Context {
  schemaIndex: Schema.Index
  config: {
    returnMode: ReturnModeType
  }
}

export const rootTypeSelectionSet = (
  context: Context,
  schemaObject: Schema.Object$2,
  ss: GraphQLObjectSelection,
  name?: string,
) => {
  return `${lowerCaseFirstLetter(schemaObject.fields.__typename.type.type)} ${name ?? ``} { ${
    selectionSet(context, schemaObject, ss)
  } }`
}

const directiveArgs = (config: object) => {
  return Object.entries(config).filter(([_, v]) => v !== undefined).map(([k, v]) => {
    return `${k}: ${JSON.stringify(v)}`
  }).join(`, `)
}

const resolveDirectives = (ss: Indicator | SS) => {
  if (isIndicator(ss)) return ``
  const { $include, $skip, $defer, $stream } = ss

  let directives = ``

  if ($stream !== undefined) {
    const config = {
      if: typeof $stream === `boolean` ? $stream : $stream.if === undefined ? true : $stream.if,
      label: typeof $stream === `boolean` ? undefined : $stream.label,
      initialCount: typeof $stream === `boolean` ? undefined : $stream.initialCount,
    }
    directives += `@stream(${directiveArgs(config)})`
  }

  if ($defer !== undefined) {
    const config = {
      if: typeof $defer === `boolean` ? $defer : $defer.if === undefined ? true : $defer.if,
      label: typeof $defer === `boolean` ? undefined : $defer.label,
    }
    directives += `@defer(${directiveArgs(config)})`
  }

  if ($include !== undefined) {
    directives += `@include(if: ${
      String(typeof $include === `boolean` ? $include : $include.if === undefined ? true : $include.if)
    })`
  }

  if ($skip !== undefined) {
    directives += `@skip(if: ${String(typeof $skip === `boolean` ? $skip : $skip.if === undefined ? true : $skip.if)})`
  }

  return directives
}

const resolveArgs = (schemaField: Schema.SomeField, ss: Indicator | SS) => {
  if (isIndicator(ss)) return ``
  const { $ } = ss
  let args = ``
  if ($ !== undefined) {
    const schemaArgs = schemaField.args
    if (!schemaArgs) throw new Error(`Field has no args`)

    const entries = Object.entries($)
    args = entries.length === 0 ? `` : `(${
      entries.map(([argName, v]) => {
        const schemaArg = schemaArgs.fields[argName] as Schema.Input.Any | undefined // eslint-disable-line
        if (!schemaArg) throw new Error(`Arg ${argName} not found in schema field`)
        if (schemaArg.kind === `Enum`) {
          return `${argName}: ${String(v)}`
        } else {
          // todo if enum, do not quote, requires schema index
          return `${argName}: ${JSON.stringify(v)}`
        }
      }).join(`, `)
    })`
  }
  return args
}
const pruneNonSelections = (ss: SS) => {
  const entries = Object.entries(ss)
  const selectEntries = entries.filter(_ => !_[0].startsWith(`$`))
  return Object.fromEntries(selectEntries)
}

const indicatorOrSelectionSet = (
  context: Context,
  schemaField: Schema.SomeField,
  ss: null | Indicator | SS,
): string => {
  if (ss === null) return `null` // todo test this case
  if (isIndicator(ss)) return ``

  const entries = Object.entries(ss)
  const selectEntries = entries.filter(_ => !_[0].startsWith(`$`))
  const directives = resolveDirectives(ss)
  const args = resolveArgs(schemaField, ss)

  if (selectEntries.length === 0) {
    return `${args} ${directives}`
  }

  const selection = Object.fromEntries(selectEntries) as GraphQLObjectSelection

  // eslint-disable-next-line
  // @ts-ignore ID error
  const schemaNamedOutputType = Schema.Output.unwrapToNamed(schemaField.type) as Schema.Object$2
  return `${args} ${directives} {
		${selectionSet(context, readMaybeThunk(schemaNamedOutputType), selection)}
	}`
}

export const selectionSet = (
  context: Context,
  schemaItem: Schema.Object$2 | Schema.Union | Schema.Interface,
  ss: Indicator | SS,
): string => {
  // todo optimize by doing single loop
  const applicableSelections = Object.entries(ss).filter(([_, ss]) => isPositiveIndicator(ss)) as [
    string,
    SS | Indicator,
  ][]
  switch (schemaItem.kind) {
    case `Object`: {
      const rootTypeName = (RootTypeName as Record<string, RootTypeName>)[schemaItem.fields.__typename.type.type]
        ?? null
      return applicableSelections.map(([fieldExpression, ss]) => {
        const fieldName = parseFieldName(fieldExpression)
        const schemaField = schemaItem.fields[fieldName.actual]
        if (!schemaField) throw new Error(`Field ${fieldExpression} not found in schema object`)
        // dprint-ignore
        if (rootTypeName&&context.config.returnMode===`successData`&&context.schemaIndex.error.rootResultFields[rootTypeName][fieldName.actual]) {
          (ss as Record<string, boolean>)[`__typename`] = true
        }
        return `${resolveFragment(resolveAlias(fieldExpression))} ${indicatorOrSelectionSet(context, schemaField, ss)}`
      }).join(`\n`) + `\n`
    }
    case `Interface`: {
      return applicableSelections.map(([fieldExpression, ss]) => {
        const fieldItem = parseFieldItem(fieldExpression)
        switch (fieldItem._tag) {
          case `FieldName`: {
            if (fieldItem.actual === `__typename`) {
              return `${renderFieldName(fieldItem)} ${resolveDirectives(ss)}`
            }
            const schemaField = schemaItem.fields[fieldItem.actual]
            if (!schemaField) throw new Error(`Field ${fieldExpression} not found in schema object`)
            // dprint-ignore
            return `${resolveFragment(resolveAlias(fieldExpression))} ${ indicatorOrSelectionSet(context, schemaField, ss) }`
          }
          case `FieldOn`: {
            const schemaObject = context.schemaIndex[`objects`][fieldItem.typeOrFragmentName]
            if (!schemaObject) throw new Error(`Fragment ${fieldItem.typeOrFragmentName} not found in schema`)
            return `${renderOn(fieldItem)} ${resolveDirectives(ss)} { ${selectionSet(context, schemaObject, ss)} }`
          }
          default: {
            throw new Error(`Unknown field item tag`)
          }
        }
      }).join(`\n`) + `\n`
    }
    case `Union`: {
      return applicableSelections.map(([fieldExpression, ss]) => {
        const fieldItem = parseFieldItem(fieldExpression)
        switch (fieldItem._tag) {
          case `FieldName`: {
            if (fieldItem.actual === `__typename`) {
              return `${renderFieldName(fieldItem)} ${resolveDirectives(ss)}`
            }
            // todo
            throw new Error(`todo resolve common interface fields from unions`)
          }
          case `FieldOn`: {
            const schemaObject = context.schemaIndex[`objects`][fieldItem.typeOrFragmentName]
            if (!schemaObject) throw new Error(`Fragment ${fieldItem.typeOrFragmentName} not found in schema`)
            // if (isIndicator(ss)) throw new Error(`Union field must have selection set`)
            return `${renderOn(fieldItem)} ${resolveDirectives(ss)} { ${
              // @ts-expect-error fixme
              selectionSet(context, schemaObject, pruneNonSelections(ss))} }`
          }
          default: {
            throw new Error(`Unknown field item tag`)
          }
        }
      }).join(`\n`) + `\n`
    }
    default:
      throw new Error(`Unknown schema item kind`)
  }
}

type FieldItem = FieldOn | FieldName

const parseFieldItem = (field: string): FieldItem => {
  const on = parseOnExpression(field)
  if (on) return on
  return parseFieldName(field)
}

interface FieldOn {
  _tag: 'FieldOn'
  typeOrFragmentName: string
}

const parseOnExpression = (field: string): null | FieldOn => {
  const match = field.match(fragmentPattern)
  if (match?.groups) {
    return {
      _tag: `FieldOn`,
      typeOrFragmentName: match.groups[`name`]!,
    }
  }
  return null
}

const renderOn = (on: FieldOn) => {
  return `...on ${on.typeOrFragmentName}`
}

// todo use a given schema to ensure that field is actually a fragment and not just happened to be using pattern onX
const resolveFragment = (field: string) => {
  const match = field.match(fragmentPattern)
  if (match?.groups) {
    return `...on ${match.groups[`name`]!}`
  }
  return field
}

interface FieldName {
  _tag: 'FieldName'
  actual: string
  alias: string | null
}
// todo use a given schema to ensure that field is actually a fragment and not just happened to be using pattern onX
const parseFieldName = (field: string): FieldName => {
  const match = field.match(aliasPattern)
  if (match?.groups) {
    return {
      _tag: `FieldName`,
      actual: match.groups[`actual`]!,
      alias: match.groups[`alias`]!,
    }
  }
  return {
    _tag: `FieldName`,
    actual: field,
    alias: null,
  }
}

const renderFieldName = (fieldName: FieldName) => {
  if (fieldName.alias) {
    return `${fieldName.actual}: ${fieldName.alias}`
  } else {
    return fieldName.actual
  }
}

// todo use a given schema to ensure that field is actually a fragment and not just happened to be using pattern onX
const resolveAlias = (field: string) => {
  const match = field.match(aliasPattern)
  if (match?.groups) {
    return `${match.groups[`actual`]!}: ${match.groups[`alias`]!}`
  }
  return field
}

const isIndicator = (v: any): v is Indicator => {
  return String(v) in indicator
}

const isPositiveIndicator = (v: any): v is SelectionSet.ClientIndicatorPositive => {
  return !(String(v) in negativeIndicator)
}

const negativeIndicator = {
  '0': 0,
  'false': false,
  'undefined': undefined,
}

const positiveIndicator = {
  '1': 1,
  'true': true,
}

const indicator = {
  ...negativeIndicator,
  ...positiveIndicator,
}
