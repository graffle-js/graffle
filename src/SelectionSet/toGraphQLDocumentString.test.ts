import { parse, print } from 'graphql'
import { expect, test } from 'vitest'
import type * as Schema from '../../tests/builder/_/schema.js'
import type { SelectionSet } from './__.js'
import { toGraphQLDocumentString } from './toGraphQLDocumentString.js'

type Q = SelectionSet.Query<Schema.$.Index>
const s = (selectionSet: Q) => selectionSet

test.each([
  s({ string: true }),
  s({ string: 1 }),
  // s({ string: false }), // todo should be static error
  s({ id: true, string: false }),
  s({ id: true, string: 0 }),
  s({ id: true, string: undefined }),
  s({ object: { id: true } }),
  s({ objectNested: { object: { string: true, id: true, int: false } } }),
  // $include
  s({ object: { $include: true, id: true } }),
  s({ object: { $include: false, id: true } }),
  s({ object: { $include: undefined, id: true } }),
  s({ object: { $include: { if: true }, id: true } }),
  s({ object: { $include: { if: false }, id: true } }),
  s({ object: { $include: { if: undefined }, id: true } }),
  s({ object: { $include: {}, id: true } }),
  // $skip
  s({ object: { $skip: true, id: true } }),
  s({ object: { $skip: false, id: true } }),
  s({ object: { $skip: undefined, id: true } }),
  s({ object: { $skip: { if: true }, id: true } }),
  s({ object: { $skip: { if: false }, id: true } }),
  s({ object: { $skip: { if: undefined }, id: true } }),
  s({ object: { $skip: {}, id: true } }),
])(`Query`, (ss) => {
  const graphqlDocumentString = toGraphQLDocumentString(ss)
  // Should parse, ensures is syntactically valid graphql document.
  const document = parse(graphqlDocumentString)
  const graphqlDocumentStringFormatted = print(document)
  const beforeAfter = `\n` + JSON.stringify(ss, null, 2)
    + `\n--------------\n` + graphqlDocumentStringFormatted + `\n`
  expect(beforeAfter).toMatchSnapshot()
})
