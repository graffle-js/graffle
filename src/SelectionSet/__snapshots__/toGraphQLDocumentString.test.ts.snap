// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Query 1`] = `
"
{
  "string": true
}
--------------
{
  string
}
"
`;

exports[`Query 2`] = `
"
{
  "string": 1
}
--------------
{
  string
}
"
`;

exports[`Query 3`] = `
"
{
  "id": true,
  "string": false
}
--------------
{
  id
}
"
`;

exports[`Query 4`] = `
"
{
  "id": true,
  "string": 0
}
--------------
{
  id
}
"
`;

exports[`Query 5`] = `
"
{
  "id": true
}
--------------
{
  id
}
"
`;

exports[`Query 6`] = `
"
{
  "object": {
    "id": true
  }
}
--------------
{
  object {
    id
  }
}
"
`;

exports[`Query 7`] = `
"
{
  "objectNested": {
    "object": {
      "string": true,
      "id": true,
      "int": false
    }
  }
}
--------------
{
  objectNested {
    object {
      string
      id
    }
  }
}
"
`;

exports[`Query 8`] = `
"
{
  "object": {
    "$include": true,
    "id": true
  }
}
--------------
{
  object @include(if: true) {
    id
  }
}
"
`;

exports[`Query 9`] = `
"
{
  "object": {
    "$include": false,
    "id": true
  }
}
--------------
{
  object @include(if: false) {
    id
  }
}
"
`;

exports[`Query 10`] = `
"
{
  "object": {
    "id": true
  }
}
--------------
{
  object {
    id
  }
}
"
`;

exports[`Query 11`] = `
"
{
  "object": {
    "$include": {
      "if": true
    },
    "id": true
  }
}
--------------
{
  object @include(if: true) {
    id
  }
}
"
`;

exports[`Query 12`] = `
"
{
  "object": {
    "$include": {
      "if": false
    },
    "id": true
  }
}
--------------
{
  object @include(if: false) {
    id
  }
}
"
`;

exports[`Query 13`] = `
"
{
  "object": {
    "$include": {},
    "id": true
  }
}
--------------
{
  object @include(if: true) {
    id
  }
}
"
`;

exports[`Query 14`] = `
"
{
  "object": {
    "$include": {},
    "id": true
  }
}
--------------
{
  object @include(if: true) {
    id
  }
}
"
`;

exports[`Query 15`] = `
"
{
  "object": {
    "$skip": true,
    "id": true
  }
}
--------------
{
  object @skip(if: true) {
    id
  }
}
"
`;

exports[`Query 16`] = `
"
{
  "object": {
    "$skip": false,
    "id": true
  }
}
--------------
{
  object @skip(if: false) {
    id
  }
}
"
`;

exports[`Query 17`] = `
"
{
  "object": {
    "id": true
  }
}
--------------
{
  object {
    id
  }
}
"
`;

exports[`Query 18`] = `
"
{
  "object": {
    "$skip": {
      "if": true
    },
    "id": true
  }
}
--------------
{
  object @skip(if: true) {
    id
  }
}
"
`;

exports[`Query 19`] = `
"
{
  "object": {
    "$skip": {
      "if": false
    },
    "id": true
  }
}
--------------
{
  object @skip(if: false) {
    id
  }
}
"
`;

exports[`Query 20`] = `
"
{
  "object": {
    "$skip": {},
    "id": true
  }
}
--------------
{
  object @skip(if: true) {
    id
  }
}
"
`;

exports[`Query 21`] = `
"
{
  "object": {
    "$skip": {},
    "id": true
  }
}
--------------
{
  object @skip(if: true) {
    id
  }
}
"
`;
